spec/requests/well_known/oauth_metadata_spec.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe 'The /.well-known/oauth-authorization-server request' do
+  it 'returns http success with valid JSON response' do
+    get '/.well-known/oauth-authorization-server'
+
+    expect(response)
+      .to have_http_status(200)
+      .and have_attributes(
+        media_type: 'application/json'
+      )
+
+    expect(body_as_json).to match(
+      a_hash_including(
```

Comment:
For whatever reason, I couldn't add tests for the important URLs that are contained within the response, because the tests and controller seemed to have different ideas of the servers URL (and I couldn't find other cases where this had been solved)

Reply:
This doesn't seem to work? I added the `host!` line above the `get` and nothing changed? The failure looks like:

```
Failure/Error:
       expect(body_as_json).to match(
         a_hash_including(
           authorization_endpoint: oauth_authorization_url,
           # FIXME: Include tests for the important URLs (for some reason routing
           # was generating mismatching URLs between the serializer and the tests)
           scopes_supported: Doorkeeper.configuration.scopes.map(&:to_s)
         )
       )
     
       expected {:authorization_endpoint=>"https://mastodon.lan:3000/oauth/authorize", :grant_types_supported=>["auth...oauth/token", :token_endpoint_auth_methods_supported=>["client_secret_basic", "client_secret_post"]} to match (a hash including {:authorization_endpoint => "http://example.com/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       Diff:
       @@ -1,10 +1,19 @@
       -(a hash including {:authorization_endpoint => "http://example.com/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       +:authorization_endpoint => "https://mastodon.lan:3000/oauth/authorize",
       +:grant_types_supported => ["authorization_code", "password", "client_credentials"],
       +:issuer => "https://mastodon.lan:3000/",
       +:registration_endpoint => "https://mastodon.lan:3000/api/v1/apps",
       +:response_modes_supported => ["query", "fragment", "form_post"],
       +:response_types_supported => ["code"],
       +:revocation_endpoint => "https://mastodon.lan:3000/oauth/revoke",
       +:scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"],
       +:token_endpoint => "https://mastodon.lan:3000/oauth/token",
       +:token_endpoint_auth_methods_supported => ["client_secret_basic", "client_secret_post"],
       
     # ./spec/requests/well_known/oauth_metadata_spec.rb:18:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:105:in `block (2 levels) in <top (required)>'
```

Notice how the expected object has `https://mastodon.lan:3000/` in it instead of the test host of `https://example.com`?

Reply:
Yeah ... do you have `mastodon.lan` set as the `LOCAL_DOMAIN` env var? I believe the server-side is going to use whatever hostname is in the request (so if we leave default, it will use `www.example.com`; if we override with `host!` it should use that) ... but the spec itself is just running the `_url` helper which is going to use whatever it's configured for to build the expected value here (which I believe will be whatever LOCAL_DOMAIN is set to).

All that said - does something like this commit do what you wanted here? - https://github.com/mastodon/mastodon/commit/46f712c1de37b9d39fa36e7badfc149db15591e0

Reply:
Unfortunately not: 
<img width="1162" alt="Screenshot 2024-02-13 at 23 29 45" src="https://github.com/mastodon/mastodon/assets/30827/f329f877-6a2e-4d3d-a7ba-86388358b3ef">

I'm guessing the controller is somehow producing mastodon.lan URLs when it should be using whatever the tests should use?

Reply:
The `cb6e6126.ngrok.io` value is the default LOCAL_DOMAIN value from `.env.test` which I would expect to be in place on the server side unless it was otherwise overridden or configured.

Where is that `mastodon.lan` value coming from?

Reply:
It's odd to me that even with both of these set, you'd wind up with a scenario where the request-spec perspective of what the var is is not the same as the server-response perspective, since they are both initially loaded the same way and you'd expect dotenv to resolve to one value of that env var and for it to be the same for both sides.

Do you have spring running? Do you have any other "interesting" env/process stuff?

What output do you see from `bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'`

Reply:
```shell
$ bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'
mastodon.lan

$ RAILS_ENV=test bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'
cb6e6126.ngrok.io

$ which rspec
rspec: aliased to RAILS_ENV=test bundle exec rspec
```

Reply:
Another idea -- do you have a value for `WEB_DOMAIN` set in your `.env`?

Reply:
In my `.env`, yes — probably don't need it though. Removing it fixed the issue partially though.

```
 Failure/Error:
       expect(body_as_json).to match(
         a_hash_including(
           authorization_endpoint: oauth_authorization_url,
           # FIXME: Include tests for the important URLs (for some reason routing
           # was generating mismatching URLs between the serializer and the tests)
           scopes_supported: Doorkeeper.configuration.scopes.map(&:to_s)
         )
       )
     
       expected {:authorization_endpoint=>"https://cb6e6126.ngrok.io/oauth/authorize", :grant_types_supported=>["auth...oauth/token", :token_endpoint_auth_methods_supported=>["client_secret_basic", "client_secret_post"]} to match (a hash including {:authorization_endpoint => "http://cb6e6126.ngrok.io/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       Diff:
       @@ -1,10 +1,19 @@
       -(a hash including {:authorization_endpoint => "http://cb6e6126.ngrok.io/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       +:authorization_endpoint => "https://cb6e6126.ngrok.io/oauth/authorize",
       +:grant_types_supported => ["authorization_code", "password", "client_credentials"],
       +:issuer => "https://cb6e6126.ngrok.io/",
       +:registration_endpoint => "https://cb6e6126.ngrok.io/api/v1/apps",
       +:response_modes_supported => ["query", "fragment", "form_post"],
       +:response_types_supported => ["code"],
       +:revocation_endpoint => "https://cb6e6126.ngrok.io/oauth/revoke",
       +:scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"],
       +:token_endpoint => "https://cb6e6126.ngrok.io/oauth/token",
       +:token_endpoint_auth_methods_supported => ["client_secret_basic", "client_secret_post"],
```

I'm probably using the wrong matcher now.

Reply:
Excellent, re: hostnames. I think what was happening is:

- The `.env.test` has a value for `LOCAL_DOMAIN` but none for `WEB_DOMAIN`, so from the controller/server/response side of things, that `LOCAL_DOMAIN` is what will be used to generate these URLs
- If you have an `.env` (ie, NOT `.env.development`) that's going to load as well in test env, and even though the `.test` one will take priority over LOCAL_DOMAIN from your `.env`, the `WEB_DOMAIN` from the `.env` will get set, and then instruct the server/response side to use that.
- On the spec side, if you don't provide a host it will use `www.example.com` and give us your initial failures, and then if you use the value I suggested from `ENV[LOCAL_DOMAIN]` it will set that, but the server will be using WEB_DOMAIN. Presumably setting it explicitly to `WEB_DOMAIN` in the spec would have worked.

All that said - I think your remaining error there is around `http` vs `https` -- see my linked commit diff for args to send the url helper.

I have a config change idea for how we might have avoid this scenario, may submit separate PR.

Reply:
@mjankowski okay, have it working, but I'm still not really happy about having to pass the protocol option to `url_for` helpers, nor the `host!` thing, feels like this should be handled in `spec_helper.rb` or something?

Or is this just the rare case where we're actually asserting complete URLs? (That doesn't sound like a correct statement)

Reply:
Yeah - I think just setting it explicitly in the rspec config makes more sense. Opened this along those lines: https://github.com/mastodon/mastodon/pull/29205

Reply:
The https portion is similar -- we have config in `.env.test` which enables HTTPS in test env: `LOCAL_HTTPS=true` ... this configures the url generators to be built as https urls. I assume this is set this way to have the test server env be as close to a typical production setup as we can get, but it does have the occassional awkward edge case like this where you need to be a little more explicit in a spec.